sanic/http.py

```diff
@@ -0,0 +1,406 @@
+        return self.stage in (Stage.HANDLER, Stage.RESPONSE)
+
+    async def http1(self):
+        """HTTP 1.1 connection handler"""
+        while True:  # As long as connection stays keep-alive
+            try:
+                # Receive and handle a request
+                self.stage = Stage.REQUEST
+                self.response_func = self.http1_response_header
+                await self.http1_request_header()
+                await self.protocol.request_handler(self.request)
+                # Handler finished, response should've been sent
+                if self.stage is Stage.HANDLER:
+                    raise ServerError("Handler produced no response")
+                if self.stage is Stage.RESPONSE:
+                    await self.response.send(end_stream=True)
+            except CancelledError:
+                # Write an appropriate response before exiting
+                e = self.exception or ServiceUnavailable(f"Cancelled")
+                self.exception = None
+                self.keep_alive = False
+                await self.error_response(e)
+            except Exception as e:
+                # Write an error response
+                await self.error_response(e)
+            # Try to consume any remaining request body
+            if self.request_body:
+                if self.response and 200 <= self.response.status < 300:
+                    logger.error(f"{self.request} body not consumed.")
+                try:
+                    async for _ in self:
+                        pass
+                except PayloadTooLarge:
+                    self.keep_alive = False
+            # Exit and disconnect if no more requests can be taken
+            if self.stage is not Stage.IDLE or not self.keep_alive:
+                break
+            # Wait for next request
+            if not self.recv_buffer:
+                await self._receive_more()
+
+    async def http1_request_header(self):
+        """Receive and parse request header into self.request."""
+        # Receive until full header is in buffer
+        buf = self.recv_buffer
+        pos = 0
+        while True:
+            pos = buf.find(b"\r\n\r\n", pos)
+            if pos != -1:
+                break
+            pos = max(0, len(buf) - 3)
+            if pos >= self.request_max_size:
+                break
+            await self._receive_more()
+        if pos >= self.request_max_size:
+            raise PayloadTooLarge("Request header exceeds the size limit")
+        # Parse header content
+        try:
+            raw_headers = buf[:pos].decode(errors="surrogateescape")
+            reqline, *raw_headers = raw_headers.split("\r\n")
+            method, self.url, protocol = reqline.split(" ")
+            if protocol == "HTTP/1.1":
+                self.keep_alive = True
+            elif protocol == "HTTP/1.0":
+                self.keep_alive = False
+            else:
+                raise Exception
```

Comment:
Probably should place an explicit exception here.

Reply:
This triggers a Bad Request response a few lines down in the same function. I've added a comment to clarify this.